src/validation/rules/NoIntrospectionFieldsRule.js

```diff
@@ -0,0 +1,35 @@
+// @flow strict
+
+import { GraphQLError } from '../../error/GraphQLError';
+
+import { type FieldNode } from '../../language/ast';
+import { type ASTVisitor } from '../../language/visitor';
+
+import { type ValidationContext } from '../ValidationContext';
+
+/**
+ * No introspection fields
+ *
+ * A GraphQL document is only valid if all fields selected are not
+ * fields that return an introspection type. Note: This rule is not
+ * part of the Validation section of the GraphQL Specification.
+ */
+export function NoIntrospectionFieldsRule(
+  context: ValidationContext,
+): ASTVisitor {
+  return {
+    Field(node: FieldNode) {
+      if (
+        context.getSchema().getQueryType() === context.getParentType() &&
+        (node.name.value === '__schema' || node.name.value === '__type')
```

Comment:
It's better to use `getFieldDef` and compare it against constants from `src/type/instospection`.

Reply:
that way you will have single check

Reply:
> How about `BanIntrospectionQueriesRule`?

That makes sense. Would `NoIntrospectionQueriesRule` be more inline with the naming of the existing rules?

Reply:
So a name like `NoFieldsWithIntrospectionTypesRule` would maybe be more accurate.

Reply:
Technically you right. But it's better to use names that improve discoverability/readability than a name 100% correct. 
For example, `SDL` is an incorrect abbreviation since specification defines `Type System Definition Language` but we use `SDL` anyway since it what community settled on.

I think the spirit of this rule is that we forbidding "introspection queries" so it's better to reflect it a name even through `{ __typename }` is also introspection query.